### PR TITLE
Feature/filter blueprint ouput

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Acceptance test rework for filter API 

Filter blueprint resource can be updated and submitted as many times as
the user wants. On submission a filter output resource is created and
will be updated with download urls by the backend export process. When
submitting a filter blueprint, the response body will return a link to
the filter output resource.

### How to review

Check logic makes sense and follows spec. Make sure all acceptance tests pass and do not leave data in mongo db. To successfully run filterAPI acceptance tests, do the following:

1) Make sure mongo db instance is running, can use mongod & (if this doesn;t work you probably need to install mongo)
2) git clone dp-dataset-api (`cmd-develop` branch will work)
3) Run dataset API (make debug)
4) git clone dp-filter-api
5) git checkout `feature/create-filter-output-resource`
6) Run filter API (make debug)
7) Run acceptance tests:
  a) cd filterAPI directory in dp-api-tests repo.
  b) go test ./...

### Who can review

Team A or B

Notes:

Pull request needs to go in with the following PR's:

FE services:
* go-ns (filter-api client): https://github.com/ONSdigital/go-ns/pull/38#pullrequestreview-74029213
* frontend-filter-dataset-controller: https://github.com/ONSdigital/dp-frontend-filter-dataset-controller/pull/37/files
* frontend-dataset-controller: https://github.com/ONSdigital/dp-frontend-dataset-controller/pull/19#pullrequestreview-74030419
* frontend-router: https://github.com/ONSdigital/dp-frontend-router/pull/47/files

BE services:
* dp-filter-api: https://github.com/ONSdigital/dp-filter-api/pull/38
* dp-dataset-exporter: https://github.com/ONSdigital/dp-dataset-exporter/pull/13
* dp-dataset-exporter-xlsx: ONSdigital/dp-dataset-exporter-xlsx#11

### CODE REVIEW

File: filterAPI/get_filter_output_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+				response.Value("dimensions").Array().Element(0).Object().Value("options").Equal([]string{"27"})
+				response.Value("downloads").Object().Value("csv").Object().Value("url").Equal("s3-csv-location")
+				response.Value("downloads").Object().Value("csv").Object().Value("size").Equal("12mb")
+				response.Value("downloads").Object().Value("json").Object().Value("url").Equal("s3-json-location")
+				response.Value("downloads").Object().Value("json").Object().Value("size").Equal("6mb")
+				response.Value("downloads").Object().Value("xls").Object().Value("url").Equal("s3-xls-location")
+				response.Value("downloads").Object().Value("xls").Object().Value("size").Equal("24mb")
+				response.Value("filter_id").Equal(filterOutputID)
+				response.Value("instance_id").Equal(instanceID)
+				response.Value("links").Object().Value("self").Object().Value("href").String().Match("(.+)/filters/" + filterOutputID + "$")
```
Comment (gedge) [inline filterAPI/get_filter_output_test.go]: FYI, the `(.+)` is largely redundant here (the parentheses capture something you don't use, for starters) and same for most uses of `Match()` - here and other files - you could remove `(.+)` or replace with `.` (if you need to test for "there's something non-empty, here")

File: filterAPI/get_filter_output_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+	filterID := uuid.NewV4().String()
+	filterOutputID := uuid.NewV4().String()
+	instanceID := uuid.NewV4().String()
+
+	filterAPI := httpexpect.New(t, cfg.FilterAPIURL)
+
+	Convey("Given an existing filter output with downloads", t, func() {
+
+		update := GetValidFilterOutputWithMultipleDimensions(cfg.FilterAPIURL, filterID, instanceID, filterOutputID)
```
Comment (gedge) [inline filterAPI/get_filter_output_test.go]: this function name is not clear that it's returning a DB/bson object

File: filterAPI/initialise.go [hunk 3]
```diff
@@ -13,6 +13,10 @@ var cfg *config.Config
 const (
 	database   = "filters"
 	collection = "filters"
+
+	internalToken          = "Internal-Token"
```
Comment (gedge) [inline filterAPI/initialise.go]: `internalTokenHeader` would be more accurate

File: filterAPI/post_dimension_option_test.go [hunk 4]
```diff
@@ -105,27 +105,27 @@ func TestFailureToPostDimensionOptions(t *testing.T) {
...
-					Expect().Status(http.StatusForbidden).Body().Contains("Forbidden, the filter job has been locked as it has been submitted to be processed\n")
-			})
-		})
-
-		if err := mongo.Teardown(database, collection, "_id", filterID); err != nil {
-			log.ErrorC("Unable to remove test data from mongo db", err, nil)
-			os.Exit(1)
-		}
-	})
+	// TODO Remove test, for now it'll be useful to keep for filter output tests
```
Comment (gedge) [inline filterAPI/post_dimension_option_test.go]: does that not now exist?

File: filterAPI/post_dimension_test.go [hunk 5]
```diff
@@ -121,32 +118,33 @@ func TestFailureToPostDimension(t *testing.T) {
...
-					log.ErrorC("Unable to setup test data", err, nil)
-					os.Exit(1)
-				}
-
-				filterAPI.POST("/filters/{filter_job_id}/dimensions/Residence Type", filterJobID).
-					WithBytes([]byte(GetValidPOSTAddDimensionToFilterJobJSON())).
-					Expect().Status(http.StatusForbidden).Body().Contains("Forbidden, the filter job has been locked as it has been submitted to be processed\n")
-
-			})
+			// TODO Remove test once filter output tests have been written
```
Comment (gedge) [inline filterAPI/post_dimension_test.go]: as above

File: testDataSetup/mongo/mongo.go [hunk 6]
```diff
@@ -114,34 +114,35 @@ func SetupMany(d *ManyDocs) error {
...
-type FilterJob struct {
-	InstanceID       string      `bson:"instance_id"          json:"instance_id"`
-	DimensionListURL string      `bson:"dimension_list_url"   json:"dimension_list_url,omitempty"`
-	Dimensions       []Dimension `bson:"dimensions,omitempty" json:"dimensions,omitempty"`
-	Downloads        *Downloads  `bson:"downloads,omitempty" json:"downloads,omitempty"`
-	Events           *Events     `bson:"events,omitempty"    json:"events,omitempty"`
-	FilterID         string      `bson:"filter_job_id"        json:"filter_job_id,omitempty"`
-	State            string      `bson:"state"                json:"state,omitempty"`
-	Links            LinkMap     `bson:"links"                json:"links,omitempty"`
+// Filter represents a structure for a filter job
```
Comment (gedge) [inline testDataSetup/mongo/mongo.go]: surely, now `// Filter represents a structure for a filter blueprint or output`

**Review by gedge — CHANGES_REQUESTED** [review]
nothing major, so far

File: filterAPI/post_dimension_option_test.go (hunk 4 above)
Comment (nshumoogum) [reply filterAPI/post_dimension_option_test.go]: Yes

File: filterAPI/post_dimension_test.go (hunk 5 above)
Comment (nshumoogum) [reply filterAPI/post_dimension_test.go]: Yes

File: testDataSetup/mongo/mongo.go (hunk 6 above)
Comment (nshumoogum) [reply testDataSetup/mongo/mongo.go]: 👍

**Review by gedge — APPROVED** [review]
